modules/extension/wmts/src/main/java/org/geotools/data/wmts/client/WMTSTileFactory.java

```diff
@@ -0,0 +1,199 @@
+import javax.measure.unit.NonSI;
+import javax.measure.unit.SI;
+import javax.measure.unit.Unit;
+
+import org.geotools.geometry.jts.ReferencedEnvelope;
+import org.geotools.tile.Tile;
+import org.geotools.tile.TileFactory;
+import org.geotools.tile.TileService;
+import org.geotools.tile.impl.ZoomLevel;
+import org.geotools.util.logging.Logging;
+import org.opengis.referencing.crs.CoordinateReferenceSystem;
+import org.opengis.referencing.cs.AxisDirection;
+import org.opengis.referencing.cs.CoordinateSystem;
+
+/**
+ * Implementation of TileFactory for WMTS
+ *
+ * @author ian
+ * @author Emanuele Tajariol (etj at geo-solutions dot it)
+ */
+public class WMTSTileFactory extends TileFactory {
+    /** pixelSizeMeters */
+    private static final double PixelSizeMeters = 0.28e-3;
+    private static final Logger LOGGER = Logging
+            .getLogger(WMTSTileFactory.class.getPackage().getName());
+
+    /**
+     *
+     */
+    public WMTSTileFactory() {
+        // TODO Auto-generated constructor stub
+    }
+
+
+    @Override
+    public Tile findTileAtCoordinate(double lon, double lat, ZoomLevel zoomLevel,
+            TileService service) {
+        WMTSZoomLevel zl = (WMTSZoomLevel) zoomLevel;
+        TileMatrix tileMatrix = ((WMTSTileService) service).getMatrixSet().getMatrices()
+                .get(zl.getZoomLevel());
+        List<TileMatrixLimits> limits = ((WMTSTileService) service).getLimits();
+        TileMatrixLimits tileMatrixLimits;
+        if (limits != null && zl.getZoomLevel() < limits.size()) {
+            tileMatrixLimits = limits.get(zl.getZoomLevel());
+        } else { // probably a REST API with no limits
+            tileMatrixLimits = new TileMatrixLimits();
+            tileMatrixLimits.setMinCol(0L);
+            tileMatrixLimits.setMinRow(0L);
+            tileMatrixLimits.setMaxCol(tileMatrix.getMatrixWidth());
+            tileMatrixLimits.setMaxRow(tileMatrix.getMatrixHeight());
+        }
+
+
+        double pixelSpan = getPixelSpan(tileMatrix);
+
+        double tileSpanY = (tileMatrix.getTileHeight() * pixelSpan);
+        double tileSpanX = (tileMatrix.getTileWidth() * pixelSpan);
+        double tileMatrixMinX;
+        double tileMatrixMaxY;
+        if (tileMatrix.getCrs().getCoordinateSystem().getAxis(0).getDirection().equals(AxisDirection.EAST)) {
+            tileMatrixMinX = tileMatrix.getTopLeft().getX();
+            tileMatrixMaxY = tileMatrix.getTopLeft().getY();
+        } else {
+            tileMatrixMaxY = tileMatrix.getTopLeft().getX();
+            tileMatrixMinX = tileMatrix.getTopLeft().getY();
+        }
+        // to compensate for floating point computation inaccuracies
+        double epsilon = 1e-6;
+        long xTile = (int) Math.floor((lon - tileMatrixMinX) / tileSpanX + epsilon);
+        long yTile = (int) Math.floor((tileMatrixMaxY - lat) / tileSpanY + epsilon);
+        // to avoid requesting out-of-range tiles
+
+        if (xTile >= tileMatrixLimits.getMaxcol())
```

Comment:
OK: This part of the code worked perfectly fine for its purposes, where input filtering is properly performed , but it was not general enough indeed.
Refactored / cleaned all that was needed to.